### PR TITLE
Task 2 - Clean repo setup with Webpack

### PR DESCRIPTION
**Pull request by davidverweij**
This PR (branched of PR #1) empties the complete repo, and set's up a new (_Typescript_) React project using `webpack` and `babel`. Using `webpack-merge`, three `webpack.config` files are used to create a `dev` environment (`npm start`) and a production build (`npm run build`).

@Dmitry-White, let me know if there are any improvements to be made!

### CODE REVIEW

File: webpack.prod.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+const { merge } = require("webpack-merge");
+const common = require("./webpack.common.config");
+
+module.exports = merge(common, {
+  mode: "production",
```
[the remark this answers is not in the document]
Comment (davidverweij) [reply webpack.prod.js:5]: That a very nice point, I've implemented it and the app (even though very empty at the moment!) can now be seen on [davidverweij.com/reactjs-practice-app/](https://www.davidverweij.com/reactjs-practice-app/)

File: package.json [hunk 2]
```diff
@@ -1,38 +1,42 @@
...
+  "version": "1.0.0",
+  "description": "A single page app for a movie DB",
+  "main": "src/index.js",
   "scripts": {
-    "start": "react-scripts start",
-    "build": "react-scripts build",
-    "test": "react-scripts test",
-    "eject": "react-scripts eject"
+    "start": "webpack serve --config webpack.dev.js",
+    "build": "webpack --config webpack.prod.js"
```
[the remark this answers is not in the document]
Comment (davidverweij) [reply package.json:8]: Got it - similar to your advice at the NodeJS MP (I have to remind myself adding these steps when setting up a project...)

File: src/components/App.tsx [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import React from "react";
```
[the remark this answers is not in the document]
Comment (davidverweij) [reply src/components/App.tsx:1]: Done, I have implemented the plugin as suggested!

**Comment by davidverweij** [comment]
Thanks @Dmitry-White! Very useful feedback and I completely agree - I will have to remind myself more about the best practices of setting up a project, such as including linting and alike.
I've adjusted the code base as suggested, let me know what you think.

File: package.json [hunk 4]
```diff
@@ -1,38 +1,62 @@
...
-    "test": "react-scripts test",
-    "eject": "react-scripts eject"
+    "start": "webpack serve --config webpack.dev.js",
+    "build": "webpack --config webpack.prod.js",
+    "deploy": "npm run build && gh-pages -d dist",
+    "eslint": "eslint .",
+    "eslint:fix": "eslint ./src --fix",
+    "pretty": "prettier --check .",
+    "pretty:fix": "prettier --write .",
+    "lint": "npm run eslint && npm run pretty",
```
Comment (Dmitry-White) [inline package.json:14]: Let's go a step further this time and add a pre-push hook with [Husky](https://typicode.github.io/husky/#/)

**Review by Dmitry-White — CHANGES_REQUESTED** [review]
Also, optionally, think about adding Docker as well

**Comment by davidverweij** [comment]
Cheers @Dmitry-White, I've added a pre-commit hook with `Husky` and set up for usage with Docker. I encountered some trade-offs with the Docker setup regarding the node_modules folder and whether to mount these to the container or not. _See the newly added changes to the README.md file with my approach_. To me, it seems illogical to depend on node_modules in the local folder, as this would render the 'custom' Docker image somewhat ineffective, but it seems a contested trade-off online. Any thoughts? Perhaps we can have a quick chat about this and your experiences with this? Cheers!

**Review by Dmitry-White — CHANGES_REQUESTED** [review]
Creating a separate volume just to map host-specific node_modules is not a good idea.
A more easy solution would be to copy all (or selected) files to the container, install modules and leave only a hot-reload stuff as a volume (which is src folder, cause everything outside of src is affecting the build and therefore requires rebuild even when not using Docker).
Screenshot 2022-04-29 at 14 22 08
Screenshot 2022-04-29 at 14 22 12

**Comment by davidverweij** [comment]
Ah cheers, good solution - thanks for that! I've changed it, and also added docker commands as `npm scripts` to streamline the setup, build and push processes. It now takes into account the package/project version so that each branch/task/PR can pull the respective docker images based on the state of the MP project.

**Review by Dmitry-White — APPROVED** [review]
Looks awesome now!

**Pull request merged**